enhancements/monitoring/multi-tenant-alerting.md

```diff
@@ -0,0 +1,573 @@
+* Specific integration in the OCP Console exposing the configuration of alert notifications.
+* Support the configuration of alert notifications for platform alerts (e.g.
+alerts originating from namespaces with the `openshift.io/cluster-monitoring: "true"`
+label).
+* Share alert receivers and routes across tenants.
+* Deploy an additional UWM Alertmanager.
+
+## Proposal
+
+We plan to leverage the `AlertmanagerConfig` custom resource definition already
+exposed by the Prometheus operator so that application owners can configure how
+and where their alert notifications should be routed.
+
+### User Stories
+
+Personas:
+* Application owners: manage a project with sufficient permissions to define monitoring resources.
+* UWM admins: manage the configuration of the UWM components (edit permissions on the `openshift-user-workload-monitoring/user-workload-monitoring-config` configmap).
+* Cluster admins: manage the configuration of the Platform monitoring components.
+
+#### Story 1
+
+As an application owner, I want to use AlertmanagerConfig custom resources so
+that Alertmanager can push alert notifications for my applications to the
+receiver of my choice.
+
+#### Story 2
+
+As an application owner, I want to use AlertmanagerConfig custom
+resources so that Alertmanager can inhibit alerts based on other alerts firing
+at the same time.
+
+#### Story 3
+
+As an application owner, I want to know if my AlertmanagerConfig custom
+resource has been reconciled on the target Alertmanager so that I am confident
+that I receive alert notifications.
+
+#### Story 4
+
+As a OpenShift cluster admin, I want to allow some of my users to
+create/update/delete AlertmanagerConfig custom resources and leverage the
+platform Alertmanager cluster so that I don't have to configure alert routing
+on their behalf.
+
+#### Story 5
+
+As an OpenShift cluster admin, I want to distinguish between platform and user
+alerts so that my Alertmanager configuration can reliably handle all platform alerts.
+
+#### Story 6
+
+As an OpenShift cluster admin, I want to exclude certain user namespaces from
+modifying the Plaform Alertmanager configuration so that I can recover in case
+of breakage or bad behavior.
+
+#### Story 7
+
+As an OpenShift cluster admin, I don't want to support AlertmanagerConfig
+resources for application owners so that the configuration of the Platform
+Alertmanager cluster is completely under my control.
+
+#### Story 8
+
+As a UWM admin, I don't want to send user alerts to the Platform Alertmanager
+cluster because these alerts are managed by an external system (off-cluster Alertmanager for
+instance).
+
+### API Extensions
+
+This enhancement proposal leverages the `AlertmanagerConfig` CRD which is
+exposed by the `coreos.monitoring.com/v1alpha1` API group and defined by the
+upstream Prometheus operator.
+
+The cluster monitoring operator deploys a `ValidatingWebhookConfiguration`
+resource to validate the `AlertmanagerConfig` resources.
```

Comment:
for what?

Reply:
What must be validated that cannot be expressed with OpenAPI ?

Reply:
Well, perhaps we are missing something and it can be done, but here is one example of an issue that we hit https://github.com/prometheus-operator/prometheus-operator/pull/4413#discussion_r752067596

Reply:
Just to close this one out, this was a single example, but having combed through, there are several examples of things we cannot validate directly in OpenAPI, at least not generated by kubebuilder annotations. So we are resigned to doing what we can in the spec and filling the gaps with the validating webhook.

Reply:
list-type set will fix prometheus operator's issue.

In general, try to avoid building APIs that cannot be validated via OpenAPI. Chance is high that those APIs won't be ideomatic anyway. There must be very strong reasons to cope with the operational complexity of a webhook.

Reply:
> list-type set will fix prometheus operator's issue.

That feels like a far-reaching assumption to be honest. For the very example [you suggest](https://github.com/prometheus-operator/prometheus-operator/pull/4413#discussion_r762766473) that it will fix, it will solve the duplicate issues, but if you look a couple of [lines down](https://github.com/prometheus-operator/prometheus-operator/pull/4413/files#diff-76b0f0103ac8d11eececdae9064b60fb4ee14845fdf6c295d48e43e00d4860b7R229), you will see that we have other issues to contend with, so it is not so straightforward.

We also have [other complex issues](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go#L126-L135) which do not fit well into OpenAPI validation.

I am all for reducing complexity where we can, but in the end, we are trying to provide a mechanism to support https://prometheus.io/docs/alerting/latest/configuration/ which is not an API or config that we can change to suit OpenAPI validation, it is us who must design the API around the nuances there, and I think it is fair to say that we have not decided to reach for the webhook without valid reasons.

The alternative is we support only the simplest of use cases (which would probably discard this enhancement), or we risk accepting bad configuration and bring down alert manager in the process.

Reply:
@sttts - I fully understand and take on board your point by the way. Right now (and in the past it seems), we have been keen to provide a 1:1 mapping between upstream config(s) and CRD and perhaps we need to rethink that approach here to simplify validation and make the API more idiomatic.

I will raise that point in the community meeting for the Prometheus operator this week.

Reply:
the JSON values make sense, I see that point. Take my comment above as a rule of thumb, not as a law :-)